pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/PinotHelixResourceManager.java

```diff
@@ -2184,6 +2190,139 @@ public PinotResourceManagerResponse toggleInstance(String instanceName, boolean
+      for (String segment : mapFields.keySet()) {
+        currentHosts.addAll(mapFields.get(segment).keySet());
+      }
+      AutoRebalanceStrategy rebalanceStrategy = new AutoRebalanceStrategy(tableNameWithType, partitions, states);
+
+      String serverTenant = TableNameBuilder.forType(tableType).tableNameWithType(tenantName);
+      List<String> instancesInClusterWithTag = _helixAdmin.getInstancesInClusterWithTag(_helixClusterName, serverTenant);
+      List<String> enabledInstancesWithTag =
+          HelixHelper.getEnabledInstancesWithTag(_helixAdmin, _helixClusterName, serverTenant);
+      LOGGER.info("Current nodes for table {}: {}", tableNameWithType, currentHosts);
+      LOGGER.info("New nodes for table {}: {}", tableNameWithType, instancesInClusterWithTag);
+      LOGGER.info("Enabled nodes for table: {}", tableNameWithType, enabledInstancesWithTag);
+      ZNRecord newZnRecord =
+          rebalanceStrategy.computePartitionAssignment(instancesInClusterWithTag, enabledInstancesWithTag,
+              mapFields, new ClusterDataCache());
+      final Map<String, Map<String, String>> newMapping = newZnRecord.getMapFields();
+      for (String segmentId : newMapping.keySet()) {
+        Map<String, String> instanceStateMap = newMapping.get(segmentId);
+        idealState.getInstanceStateMap(segmentId).clear();
+        for (String instanceId : instanceStateMap.keySet()) {
+          idealState.setPartitionState(segmentId, instanceId, instanceStateMap.get(instanceId));
+        }
+      }
+      // If we removed any entries, add them back here
+      for (Map.Entry<String, Map<String, String>> entry: removedEntries) {
+        Map<String, String> stateMap = entry.getValue();
+        for (String instanceId : stateMap.keySet()) {
+          idealState.setPartitionState(entry.getKey(), instanceId, stateMap.get(instanceId));
+        }
+      }
+    }
+    idealState.setReplicas(Integer.toString(targetNumReplicas));
+    return idealState;
+  }
+
+  // Keep only those segments that are LLC and in ONLINE state.
+  private void removeSegmentsNotBalancable(Map<String, Map<String, String>> mapFields, List<Map.Entry<String, Map<String, String>>> removedEntries) {
+    // Keep only those segments that are LLC and in ONLINE state.
+    Iterator<Map.Entry<String, Map<String, String>>> it = mapFields.entrySet().iterator();
+    while (it.hasNext()) {
+      Map.Entry<String, Map<String, String>> entry = it.next();
+      final String segmentName = entry.getKey();
+      boolean keep = false;
+      if (SegmentName.isLowLevelConsumerSegmentName(segmentName)) {
+        // Check the states. If any of the instances are ONLINE, then we keep the segment for rebalancing,
+        // Otherwise, we remove it, and add it back after helix re-balances the segments.
+        Map<String, String> stateMap = entry.getValue();
+        for (String state : stateMap.values()) {
+          if (state.equals(SegmentOnlineOfflineStateModel.ONLINE)) {
```

Comment:
In the idealstate, they should never be mixed, because they all transition to ONLINE in one atomic set of the idealstate.

Even if they are mixed, the fact that one of the replicas is ONLINE means that the others can  transition to ONLINE since the segment is available for download.